packages/dds/SchemaVersioning.md

```diff
@@ -0,0 +1,159 @@
+
+-   Keep the contents of persisted data formats out of public APIs.
+
+    Since APIs have different compatibility requirements than persisted data formats, mixing the two will result in reduced evolvability long term.
+    This avoids the complexity related to supporting reading and writing multiple formats from cluttering the APIs.
+
+-   Implement support for persisted formats in a self contained way, where the relevant code is clearly marked.
+
+    A type which is persisted has very different evolvability than one that is not.
+    For example a package that has a non-exported interface usually can do a rename refactoring for members of the interface, but if that interface is used for Json serialized objects which are persisted,
+    the change could easily break support for all existing documents and/or collaboration with all existing clients, and do so in a way that won't show up in code review since the persistance/format logic might be untouched.
+    Instead any code which defined a persisted format (including interfaces used with Json) should be clearly documented as such.
+    An easy way to do this is put persisted related details in a clearly named file (for example persisted_formats.ts),
+    and do not import anything into that file which could change and result in a changed data format (for example don't import any types to include in json serialized data).
+    If necessary duplicate types, or move the source of truth to the persisted format code.
+
+-   Factor legacy format reading code (code for reading formats old enough that writing them is not supported) into compatibility libraries which can be authored as converters.
+
+    This approach ensures that the only unbounded code growth from having a very old codebase with many legacy formats (support for reading all old formats) is organized in a way that its complexity does not increase as more formats are added, and it has no maintainability impact on other parts of the code.
+
+## Libraries
+
+This includes Fluid Framework, as well other libraries consumed by Fluid applications.
+
+-   Adding support for writing a new format can be done in a minor version, but it must not be used unless opted into by the application or by editing a document already using the new format.
+
+    When adding a new format, the APIs to opt into that format should explicitly document the oldest version in which that format is supported, so that when applications consider opting into it, they can know if it is safe for them to do so.
+
+    If the format is not yet stable (current version of it will not be supported for reading in all future versions), it must be explicitly documented as such,
```

Comment:
This is an area we totally messed up with legacy SharedTree: we iterated on the 1.1.1 format in code, with it documented as not stable for production yet. However, we didn't actually rev the version to 1.1.2 or equivalent upon committing to backwards-compatibility. This ended up resulting in corrupted documents for the ~0.01% of sessions which were active for over 3 months between the dark rollout time and enabling the write feature, since they were using client versions which *thought* they knew how to interpret 1.1.1, but it was one of the intermediate iterations of 1.1.1. Thankfully the corrupted documents could be recovered using a rollback feature at the storage layer, but without that we would have needed to implement a lot of complicated hotfixes if we wanted to recover their data.

Clearly we did not follow best practices in this example. I do think there is high value in having a documented process for iterating on a work-in-progress eventual write format in a way which allows breaking format changes before finalizing the format and might be worth writing it as an example here. E.g., if we had instead kept the version number something like "1.1.1-dev" while we didn't commit to back-compat and finalized it via "promotion" to 1.1.1, those corrupt documents would have never happened.

Reply:
fyi--my more recent thinking in this area is that it's probably just better to use numbers as versions, and while working on support for a new version have explicit guards that attempts to read/write that version fail. Ex: while working on V2, in code have a `maxProductionVersion: 1` field, create codecs which assert when encountering production version >1, add an internal-only backdoor which allows writing tests using version 2. then when ready to commit to backcompat for version 2, bump `maxProductionVersion` to 2.

Reply:
I have added additional guidance about this, which I think is even more protective than having a maxProductionVersion.